README.md
# NeoPixelEspEx

Library for WS281x and compatible LEDs for ESP8266 and ESP32. It can be used as slim standalone version or together with FastLED, <s>which currently does not work properly with `framework-arduinoespressif8266 3.0.0`</s> The current master branch fixes the issue.

## FastLED Fork

Since the current FastLED 3.5 still has issues with WS281x (ESP8266), I updated my fork [https://github.com/sascha432/FastLED](https://github.com/sascha432/FastLED)

## Features

- Support for ESP8266/GPIO16
- Option to use precaching instead of IRAM
- Support for brightness scaling
- Support for interrupts and retries if interrupted (ESP8266)
- Support for GRB, RGB, CRGB (FastLED) and other types
- Function to safely clear pixels without allocating any memory, for example during boot, restart, crash...

## Examples

Examples for PlatformIO can be found in the `examples/` folder.

## FastLED

A workaround for FastLED and v3.0.0 is to replace `FastLED.show()` with `Strip::show()`. Only brightness is supported.

```c++
CRGB pixelData[NEOPIXEL_NUM_PIXELS];
NeoPixelEx::Strip<NEOPIXEL_OUTPUT_PIN, NEOPIXEL_NUM_PIXELS, NeoPixelEx::CRGB, NeoPixelEx::TimingsWS2812, NeoPixelEx::DataWrapper<NEOPIXEL_NUM_PIXELS, NeoPixelEx::CRGB>> pixels(&pixelData);

fill_rainbow(pixelData, pixels.getNumPixels(), beat8(10, 255), 10);

pixels.show(FastLED.getBrightness());
// or
FastLED.show();
```


## Basic usage

### Legacy functions

It is recommended to use the class `NeoPixelEx::Strip` instead of `NeoPixel_espShow()`.

```c++
#include <NeoPixelEspEx.h>

#define NEOPIXEL_OUTPUT_PIN 12

void setup() {
  pinMode(NEOPIXEL_OUTPUT_PIN, OUTPUT);
  ...
}

void loop() {
  ...
  // set color of first pixel to 0x20000
  uint8_t one_pixel[3] = { 0x20, 0, 0 };
  NeoPixel_espShow(NEOPIXEL_OUTPUT_PIN, one_pixel, sizeof(one_pixel));
  ...
  // fill 8 pixels with color #220011 and display
  uint8_t pixels[8 * 3];
  NeoPixel_fill(pixels, sizeof(pixels), 0x220011);
  NeoPixel_espShow(NEOPIXEL_OUTPUT_PIN, pixels, sizeof(pixels));
  ...
  // turn 256 LEDs off without reading the data from tmp
  // brightness must be 0
  NeoPixel_espShow(NEOPIXEL_OUTPUT_PIN, nullptr, 3 * 256, 0);
}

```

## Strip class

```c++
#include <NeoPixelEspEx.h>

#define NEOPIXEL_OUTPUT_PIN 12

NeoPixelEx::Strip<NEOPIXEL_OUTPUT_PIN, 8, NeoPixelEx::GRB, NeoPixelEx::TimingsWS2812> pixels;

void setup() {
  ...
  pixels.begin();
  ...
  // set all pixels color 0x000000 and update with 0 brightness
  pixels.clear();
  ...
  // set all LEDs to 0x100010 and update with full brightness
  pixels.fill(0x100010);
  pixels.show();
  ...
}

void loop() {
  ...
  // set the color of pixel 1 and 2, and update the LEDs with brightness level 32
  pixels[0] = 0xff0000;
  pixels[1] = 0x00ff00;
  pixels.show(32);
  ...
  // turn 256 LEDs off without Strip object or allocating any memory
  // NeoPixelEx::DefaultTimings is used
  NeoPixelEx::forceClear(256);
  ...
}
```

### Changing pixel data

`pixels.data()` provides access to the underlying data object. The default is `PixelData` which is an array with some additional functions to manipulare the data.

### DataWrapper

To use existing data with the Strip class, the DataWrapper template can be used. It provides direct access to the data without any additional overhead.

```c++

uint8_t pixelData[NEOPIXEL_NUM_PIXELS * sizeof(NeoPixelEx::GRB)];
NeoPixelEx::Strip<NEOPIXEL_OUTPUT_PIN, NEOPIXEL_NUM_PIXELS, NeoPixelEx::GRB, NeoPixelEx::TimingsWS2812, DataWrapper<NEOPIXEL_NUM_PIXELS, NeoPixelEx::GRB>> pixels(&pixelData);

```

## Pins ESP8266

Usable pins for LED strips. Depending on the levelshifter, GPIO 1, 3, 9 and 10 can be used. GPIO 9 and 10 require to run the flash memory in DIO mode.

  NodeMCU | PIN | Description |
|---|---|---|
| D0 | GPIO16 | floating (or connected to reset, directly or via low value series resistor = pullup) |
| D1 | GPIO5 | floating
| D2 | GPIO4 | floating
| D3 | GPIO0 | pullup
| D4 | GPIO2 | pullup
| D5 | GPIO14 | floating
| D6 | GPIO12 | floating
| D7 | GPIO13 | floating
| D8 | GPIO15 | pulldown
| RX | GPIO3 | serial2usb converter
| TX | GPIO1 | serial2usb converter
| S2 | GPIO9 | flash memory (QIO mode)
| S3 | GPIO10 | flash memory (QIO mode)

## Based on NeoPixel_esp

This library is based on

```c++
// This is a mash-up of the Due show() code + insights from Michael Miller's
// ESP8266 work for the NeoPixelBus library: github.com/Makuna/NeoPixelBus
```
and

[https://github.com/adafruit/Adafruit_NeoPixel/blob/master/esp8266.c](https://github.com/adafruit/Adafruit_NeoPixel/blob/master/esp8266.c)



CHANGELOG.md
# Changelog

## Version 0.0.3

 - Changed name to NeoPixelEspEx

## Version 0.0.2

- Support for ESP32 RMT
- Fixed wrong color of the first pixel if CRGBType was used
- Updated timings for WS2813
- GPIO16 seems to have some issues when running with 160MHz, try another port if any issues occur
- Wrapper class for plain data pointers
- Context object for each LED strip and global context object for legacy functions
- Support for espressif8266@3.0.1, 2.6.3 and 2.3.3
- Templates for timings
- GRB, RGB and Array types
- Strip class
- Clear function that does not require to allocate any memory

## Version 0.0.1

First version based on this is a mash-up of the Due show() code + insights from Michael Miller's
ESP8266 work for the NeoPixelBus library: github.com/Makuna/NeoPixelBus and
(https://github.com/adafruit/Adafruit_NeoPixel/blob/master/esp8266.c)[https://github.com/adafruit/Adafruit_NeoPixel/blob/master/esp8266.c]



scripts/extra_script.py
import re
from os import path
import os
import click
import subprocess
from SCons.Script import ARGUMENTS
from pprint import pprint
Import("env", "projenv")

def which(name, env, flags=os.F_OK):
    result = []
    extensions = env["ENV"]["PATHEXT"].split(os.pathsep)
    if not extensions:
        extensions = [".exe", ""]

    for path in env["ENV"]["PATH"].split(os.pathsep):
        path = os.path.join(path, name)
        if os.access(path, flags):
            result.append(os.path.normpath(path))
        for ext in extensions:
            whole = path + ext
            if os.access(whole, os.X_OK):
                result.append(os.path.normpath(whole))
    return result

def disassemble(source, target, env):

    verbose = int(ARGUMENTS.get("PIOVERBOSE", 0))

    source = path.abspath(env.subst('$PIOMAINPROG'))
    target = env.subst(env.GetProjectOption('custom_disassemble_target', '$BUILD_DIR/${PROGNAME}.lst'))
    target = path.abspath(target)

    command = env.subst(env.GetProjectOption('custom_disassemble_bin', env['CC'].replace('gcc', 'objdump')))

    options = re.split(r'[\s]', env.subst(env.GetProjectOption('custom_disassemble_options', '-S -C')))

    args = [command] + options + [source, '>', target]
    if verbose:
        click.echo(' '.join(args))

    return_code = subprocess.run(args, shell=True).returncode
    if return_code != 0:
        click.secho('failed to run: %s' % ' '.join(args))
        env.Exit(1)

    click.echo('-' * click.get_terminal_size()[0])
    click.secho('Created: ', fg='yellow', nl=False)
    click.secho(target)

env.AlwaysBuild(env.Alias('disasm', [env['PIOMAINPROG']], disassemble))
env.AlwaysBuild(env.Alias('disassemble', [env['PIOMAINPROG']], disassemble))


examples/esp8266_fastled_example.cpp
/**
 * Author: sascha_lammers@gmx.de
 */

#define FASTLED_INTERNAL
#define FASTLED_ESP8266_RAW_PIN_ORDER
#include <FastLED.h>
#include "NeoPixelEspEx.h"
#include <Schedule.h>
#include <ESP8266WiFi.h>

// to test with an active WiFi connection, add SSID and password or put credentials into ../../wifi.h
// #define WIFI_SSID "SSID"
// #define WIFI_PASS "123"
#ifndef WIFI_SSID
    #include "../../wifi.h"
#endif

#define POWER_LIMIT_MW 2500

#ifndef POWER_LIMIT_MW
#define POWER_LIMIT_MW 0
#endif

CRGB pixelData[NEOPIXEL_NUM_PIXELS];
// NeoPixelEx::CRGB requires more CPU cycles than NeoPixelEx::GRB
NeoPixelEx::Strip<NEOPIXEL_OUTPUT_PIN, NEOPIXEL_NUM_PIXELS, NeoPixelEx::CRGB, NeoPixelEx::TimingsWS2812, NeoPixelEx::DataWrapper<NEOPIXEL_NUM_PIXELS, NeoPixelEx::CRGB>> pixels(&pixelData);
bool useFastLEDShow = true;
int stepSize = 4;
int bpm = 10;
int mainLoopDelay = 5;
bool dithering = false;

NeoPixelEx::Context context;

static void show()
{
    if (useFastLEDShow) {
        FastLED.show();
    }
    else {
        pixels.show(FastLED.getBrightness());
    }
}

// static void _delay(int millis)
// {
//     if (useFastLEDShow) {
//         FastLED.delay(millis);
//     }
//     else {
//         delay(millis);
//     }
// }



static void help()
{
    Serial.print(F(
        "---\n"
        "+      increase brightness\n"
        "-      decrease brightness\n"
        "*      increase step size\n"
        "/      decrease step size\n"
        "0-9    set bpm from 10-100\n"
        "p      change pattern\n"
        "d      toggle display method\n"
        "D      enable/disable dithering\n"
        "w      connect to WiFi\n"
        "W      disconnect WiFi\n"
        "\n"
        "power limit set to " __STRINGIFY(POWER_LIMIT_MW) "mW\n"
        "\n")
    );
}

int pattern = 0;

void update_pattern() {
    switch(pattern % 2) {
        case 0: {
                pixels.fill(0, 10, 0x202020);
                pixels.fill(50, 10, 0x000001);
                pixels[0] = 0xff7f00;
                pixels[1] = 0xff0000;
                pixels[10] = 0xff0000;
                pixels[15] = 0x0000ff;
                pixels[25] = 0x00ff00;

                NeoPixelEx::GRB color(0x0000ff);
                for(uint16_t i = 0; i < pixels.getNumPixels() + 10; i += 10) {
                    pixels.fill(i, 10, color);
                    color <<= 8;
                    if (i % 30 == 20) {
                        color.setBrightness(200);
                    }
                }
            }
            break;
        case 1: {
                fill_rainbow(pixelData, pixels.getNumPixels(), beat8(10, 255), 10);
            } break;
    }
}

void setup()
{
    WiFi.persistent(false);
    WiFi.setAutoConnect(false);
    WiFi.setAutoReconnect(false);
    WiFi.disconnect();

    Serial.begin(115200);
    pixels.begin();
    NeoPixelEx::forceClear(NEOPIXEL_OUTPUT_PIN, 256);

    FastLED.addLeds<WS2812, NEOPIXEL_OUTPUT_PIN, GRB>(pixelData, NEOPIXEL_NUM_PIXELS);
    #if POWER_LIMIT_MW
        FastLED.setMaxPowerInVoltsAndMilliamps(5, POWER_LIMIT_MW / 5);
    #endif

    FastLED.setBrightness(10);
    FastLED.setDither(dithering);

    FastLED.clear();
    show();
    return;

    schedule_recurrent_function_us([]() {
        help();
        return false;
    }, 5 * 1000 * 1000);

    schedule_recurrent_function_us([]() {
        Serial.printf_P(PSTR("FPS: %u\n"), FastLED.getFPS());
        return true;
    }, 5 * 1000 * 1000);

    update_pattern();

    WiFi.onEvent([](WiFiEvent_t event) {
        Serial.printf_P(PSTR("WiFi event: %u\n"), event);
        if (event == WIFI_EVENT_STAMODE_GOT_IP) {
            Serial.print(F("WiFi: connected to %"));
            Serial.print(WiFi.SSID());
            Serial.print(F(" / "));
            WiFi.localIP().printTo(Serial);
            Serial.println();
        }
        else if (event == WIFI_EVENT_STAMODE_DISCONNECTED) {
            Serial.println(F("WiFi: disconnect"));
        }
    }, WIFI_EVENT_ANY);
}

NeoPixelEx::CRGB color(0xff0000);

void loop()
{
    EVERY_N_MILLISECONDS(10) {
        update_pattern();
    }
    show();

    if (Serial.available()) {
        int ch;
        switch(ch = Serial.read()) {
            case '*':
                stepSize = std::min(32, stepSize + 1);
                Serial.printf_P(PSTR("step size %u\n"), stepSize);
                break;
            case '/':
                stepSize = std::max(1, stepSize - 1);
                Serial.printf_P(PSTR("step size: %u\n"), stepSize);
                break;
            case '+':
                FastLED.setBrightness(std::min<int>(255, FastLED.getBrightness() + stepSize));
                Serial.printf_P(PSTR("brightness: %u\n"), FastLED.getBrightness());
                break;
            case '-':
                FastLED.setBrightness(std::max<int>(0, FastLED.getBrightness() - stepSize));
                Serial.printf_P(PSTR("brightness: %u\n"), FastLED.getBrightness());
                break;
            case 'p':
                pattern++;
                break;
            case '0':
            case '1':
            case '2':
            case '3':
            case '4':
            case '5':
            case '6':
            case '7':
            case '8':
            case '9':
                ch -= '0';
                bpm = ch * 10 + 10;
                Serial.printf_P(PSTR("bpm: %u\n"), bpm);
                break;
            case 'D':
                FastLED.setDither(dithering = !dithering);
                Serial.printf_P(PSTR("dithering: %u\n"), dithering);
                break;
            case 'd':
                useFastLEDShow = !useFastLEDShow;
                Serial.printf_P(PSTR("display method: %s\n"), useFastLEDShow ? PSTR("FastLED") : PSTR("NeoPixel"));
                break;
            case 'w':
                Serial.printf_P(PSTR("WiFi status: %u\n(re)connecting WiFi...\n"), WiFi.status());
                WiFi.disconnect();
                WiFi.setAutoConnect(true);
                WiFi.setAutoReconnect(true);
                WiFi.begin(WIFI_SSID, WIFI_PASS);
                break;
            case 'W':
                WiFi.setAutoConnect(false);
                WiFi.setAutoReconnect(false);
                WiFi.disconnect();
                Serial.println(F("WiFi disconnected..."));
                break;
            default:
                help();
                break;
        }
    }
}


examples/esp8266_example_crash_handler.cpp
/**
 * Author: sascha_lammers@gmx.de
 */

//
// example of a custom crash handler that turns all LEDs off
//

#include <Arduino.h>
#include "NeoPixelEspEx.h"
#include <core_version.h>

NeoPixelEx::Strip<NEOPIXEL_OUTPUT_PIN, NEOPIXEL_NUM_PIXELS, NeoPixelEx::GRB, NEOPIXEL_CHIPSET> pixels;

extern "C" void custom_crash_callback(struct rst_info *rst_info, uint32_t stack, uint32_t stack_end);

void custom_crash_callback(struct rst_info *rst_info, uint32_t stack, uint32_t stack_end)
{
    NeoPixelEx::forceClear<decltype(pixels)::chipset_type>(pixels.kOutputPin, NEOPIXEL_NUM_PIXELS);
}

void count_down(PGM_P msg, int n)
{
    for(int i = n; i > 0; i--) {
        Serial.printf_P(PSTR("%s in %u seconds...\n"), msg, i);
        delay(1000);
    }
}

void setup()
{
    Serial.begin(115200);
    Serial.println(F("Starting..."));
    count_down(PSTR("Turning LEDs on"), 10);

    pixels.begin();
    pixels.fill(0xff00ff);
    pixels.show(16);
}

void loop()
{
    count_down(PSTR("Invoking panic()"), 10);
    panic();
}


examples/esp8266_example.cpp
/**
 * Author: sascha_lammers@gmx.de
 */

#include <Arduino.h>
// #include <core_version.h>
#include <Schedule.h>
#define FASTLED_INTERNAL
#define FASTLED_ESP8266_RAW_PIN_ORDER
#include <FastLED.h>
#include "NeoPixelEspEx.h"
#include "wiring_private.h"

#define NEOPIXEL_NUM_PIXELS_CLEAR NEOPIXEL_NUM_PIXELS + 10

using namespace NeoPixelEx;

Strip<NEOPIXEL_OUTPUT_PIN, NEOPIXEL_NUM_PIXELS, NeoPixelEx::GRB, NEOPIXEL_CHIPSET> pixels;

// example of using the DataWrapper to use any raw pointer with the Strip class
//
// uint8_t pixelData[NEOPIXEL_NUM_PIXELS * sizeof(GRB)];
// Strip<NEOPIXEL_OUTPUT_PIN, NEOPIXEL_NUM_PIXELS, GRB, NEOPIXEL_CHIPSET, DataWrapper<NEOPIXEL_NUM_PIXELS, GRB>> pixels(&pixelData);

extern bool record_serial_output;
extern String serial_output;

#if 0
#include <user_interface.h>
uart_t *_uart;
void preinit()
{
    _uart = uart_init(UART0, 115200, (int) SERIAL_8N1, (int) SERIAL_FULL, 1, 64, false);
    ::printf("preinit\n");
}
#endif

void help()
{
    Serial.printf_P(PSTR("---\npin=%u pixels=%u cpu=%u\n"), pixels.kOutputPin, pixels.getNumPixels(), (int)(F_CPU / 1000000UL));
    Serial.print(F(
        "---\n" \
        "H  turn all pixels off and halt\n" \
        "R  reset pixels and values\n" \
        "C  clear pixels\n" \
        "t  test all pixels\n" \
        "+  increase selected value by step size\n" \
        "-  decrease selected value by step size\n" \
        "*  increase step size\n" \
        "/  decrease step size\n" \
        "r  select red value\n" \
        "b  select blue value\n" \
        "g  select green value\n" \
        "B  select brightness\n" \
        "d  select delay\n" \
        "S  clear stats\n" \
        "p  next pattern (param1-param3 can be used to modify it)\n" \
        "P  prev. pattern (param1-param3 can be used to modify it)\n" \
        "1  select param 1\n" \
        "2  select param 2\n" \
        "3  select param 3\n" \
        "?  display help\n" \
        "\n" \
    ));
}

void setup()
{
#if 0
    ::printf("setup\n");
    uart_uninit(_uart);
#endif
    forceClear<pixels.kOutputPin, decltype(pixels)::chipset_type>(NEOPIXEL_NUM_PIXELS_CLEAR);

    Serial.begin(115200);
    Serial.println(F("starting..."));

    // pinMode(2,OUTPUT);
    // gdbstub_init();


    pixels.begin();

    schedule_recurrent_function_us([]() {
        help();
        return false;
    }, 5000000);

}

NeoPixelEx::GRB color(0, 0, 0xff);
uint8_t brightness = 10;
uint8_t delayMillis = 4;
uint8_t pattern = 1;
uint8_t stepSize = 1;
uint8_t param1 = 10;
uint8_t param2 = 10;
uint8_t param3 = 1;
uint8_t *valuePtr = &brightness;

static void default_params() {
    switch(pattern) {
        case 1:
            param1 = 30;
            param2 = 10;
            break;
        case 2:
            param1 = 10;
            param2 = 4;
            break;
    }
}

static PGM_P get_type_str()
{
    if (valuePtr == &brightness) {
        return PSTR("brightness");
    }
    if (valuePtr == &color.r) {
        return PSTR("red");
    }
    if (valuePtr == &color.g) {
        return PSTR("green");
    }
    if (valuePtr == &color.b) {
        return PSTR("red");
    }
    if (valuePtr == &delayMillis) {
        return PSTR("delay");
    }
    if (valuePtr == &stepSize) {
        return PSTR("step size");
    }
    if (valuePtr == &pattern) {
        return PSTR("pattern");
    }
    if (valuePtr == &param1) {
        return PSTR("param1");
    }
    if (valuePtr == &param2) {
        return PSTR("param2");
    }
    if (valuePtr == &param3) {
        return PSTR("param3");
    }
    return PSTR("?");
}

static void clear_keys() {
    while(Serial.available()) {
        Serial.read();
    }
}

static int testAll = -1;
static int delayMillisOld = -1;

void end_test(const __FlashStringHelper *msg)
{
    forceClear<pixels.kOutputPin, decltype(pixels)::chipset_type>(NEOPIXEL_NUM_PIXELS_CLEAR);
    testAll = -1;
    if (delayMillisOld != -1) {
        delayMillis = delayMillisOld;
        delayMillisOld = -1;
    }
    Serial.println(msg);
}

void loop()
{
    static int counter = 0;

    EVERY_N_MILLIS(250) {
        static bool toggle;
        toggle = !toggle;
        digitalWrite(2, toggle);
    }

    if (testAll != -1 && counter > 100) {
        counter = 0;
        using ColorType = NeoPixelEx::GRB;
        // test pixels
        static ColorType onColor;

        auto numRun = testAll / pixels.getNumPixels();
        auto currentPixelNum = testAll % pixels.getNumPixels();
        if (currentPixelNum == 0) {
            if (numRun >= 4) {
                end_test(F("finished testing"));
                return;
            }
            // change color per run
            if (testAll % pixels.getNumPixels() == 0) {
                switch(numRun) {
                    case 0:
                        onColor = ColorType(0xff, 0, 0);
                        break;
                    case 1:
                        onColor = ColorType(0, 0xff, 0);
                        break;
                    case 2:
                        onColor = ColorType(0, 0, 0xff);
                        break;
                    case 3:
                        onColor = ColorType(0xff, 0xff, 0xff);
                        break;
                }
            }
        }
        ColorType offColor = onColor.inverted();
        offColor.setBrightness(128);
        #if NEOPIXEL_HAVE_STATS
            auto &stats = pixels.getStats();
            Serial.printf_P(PSTR("testing color on=%s off=%s pixel=%u/%u run=%u dropped=%u (%.1f%%) frames=%u fps=%u delay=10\n"),
                onColor.toString().c_str(), offColor.toString().c_str(), currentPixelNum, pixels.getNumPixels(), numRun,
                stats.getAbortedFrames(),  stats.getAbortedFrames() * 100.0 / stats.getFrames(), stats.getFrames(),
                stats.getFps(), delayMillis
            );
        #else

        #endif
        pixels.fill(offColor);
        pixels[currentPixelNum] = onColor;
        pixels.show(10);
        testAll++;
    }
    else if (testAll == -1) {
        EVERY_N_MILLISECONDS(10) {
            switch(pattern % 4) {
                case 0:
                    // display pixels
                    pixels.fill(color);
                    break;
                case 1: {
                        fill_rainbow(pixels.cast<::CRGB *>(), pixels.getNumPixels(), beat8(param1, 255), param2);
                    }
                    break;
                case 2: {
                        uint32_t from = 0x00ff00;
                        uint32_t to = 0xff00ff;
                        for(auto i = 0; i < pixels.getNumPixels() && i < pixels.getNumPixels(); i += param1) {
                            fill_gradient_RGB(&pixels.cast<::CRGB *>()[i], pixels.getNumPixels() - i, from, to);
                            std::swap(from, to);
                            // auto tmp = NeoPixelEx::GRB(from);
                            // tmp <<= 4;
                            // from = tmp.toRGB();
                            auto tmp = NeoPixelEx::GRB(to);
                            tmp <<= param2;
                            to = tmp.toRGB();
                        }
                    } break;
                default:
                    pixels.fill(0);
                    break;
            }
        }
        pixels.show(brightness);
    }

    // serial input handler
    if (Serial.available()) {
        if (testAll != -1) {
            end_test(F("testing pixels aborted..."));
        }
        while(Serial.available()) {
            char ch  = Serial.read();
            switch(ch) {
                case 'M': {
                        pixels.fill(0);
                        auto &data = pixels.data();
                        auto endPtr = data.end();
                        for(auto ptr = data.begin(); ptr < endPtr; ptr += 10) {
                            *ptr = 0x000011;
                        }
                        for(auto ptr = data.begin() + 4; ptr < endPtr; ptr += 10) {
                            *ptr = 0x000044;
                        }
                        for(auto ptr = data.begin() + 9; ptr < endPtr; ptr += 10) {
                            *ptr = 0x0f0088;
                        }
                        for(auto ptr = data.begin() + 49; ptr < endPtr; ptr += 50) {
                            *ptr = 0x004400;
                        }
                        for(auto ptr = data.begin() + 99; ptr < endPtr; ptr += 100) {
                            *ptr = 0x008800;
                        }
                        pixels[pixels.size() - 1] = 0xff0000;
                        pixels.show(255);
                        delay(30000);
                        pixels.clear();
                    }
                    break;
                case 't':
                    if (delayMillisOld == -1) {
                        delayMillisOld = delayMillis;
                    }
                    delayMillis = 10;
                    Serial.println(F("testing all pixels\n"));
                    clear_keys();
                    testAll = 0;
                    counter = -1;
                    #if NEOPIXEL_HAVE_STATS
                        pixels.getStats().clear();
                    #endif
                    pixels.clear();
                    break;
                case 'p':
                    pattern++;
                    default_params();
                    break;
                case 'P':
                    pattern--;
                    default_params();
                    break;
                case '*':
                    stepSize = std::min(32, stepSize + 1);
                    break;
                case '/':
                    stepSize = std::max<int>(1, stepSize - 1);
                    break;
                case '+':
                    if (valuePtr == &pattern || valuePtr == &param1 || valuePtr == &param2 || valuePtr == &param3) {
                        *valuePtr = std::min<int>(255, *valuePtr + 1);
                    }
                    else {
                        *valuePtr = std::min<int>(255, *valuePtr + stepSize);
                    }
                    break;
                case '-':
                    if (valuePtr == &pattern || valuePtr == &param1 || valuePtr == &param2 || valuePtr == &param3) {
                        *valuePtr = std::min<int>(255, *valuePtr - 1);
                    }
                    else {
                        *valuePtr = std::max<int>(0, *valuePtr - stepSize);
                    }
                    break;
                case 'r':
                    valuePtr = &color.r;
                    break;
                case 'b':
                    valuePtr = &color.b;
                    break;
                case 'g':
                    valuePtr = &color.g;
                    break;
                case 'B':
                    valuePtr = &brightness;
                    break;
                case 'd':
                    valuePtr = &delayMillis;
                    break;
                case '1':
                    valuePtr = &param1;
                    break;
                case '2':
                    valuePtr = &param2;
                    break;
                case '3':
                    valuePtr = &param3;
                    break;
                case 'S':
                    Serial.println(F("clearing stats..."));
                    #if NEOPIXEL_HAVE_STATS
                        pixels.getStats().clear();
                    #endif
                    break;
                case 'R':
                    Serial.println(F("reset"));
                    color = NeoPixelEx::GRB(0, 0, 0xff);
                    brightness = 16;
                    delayMillis = 1;
                    valuePtr = &brightness;
                    // falltrough
                case 'C':
                    Serial.println(F("clearing all pixels..."));
                    forceClear<pixels.kOutputPin, decltype(pixels)::chipset_type>(NEOPIXEL_NUM_PIXELS_CLEAR);
                    Serial.println(F("waiting 3 seconds..."));
                    delay(3000);
                    clear_keys();
                    #if NEOPIXEL_HAVE_STATS
                        pixels.getStats().clear();
                    #endif
                    break;
                case 'H':
                    forceClear<pixels.kOutputPin, decltype(pixels)::chipset_type>(NEOPIXEL_NUM_PIXELS_CLEAR);
                    Serial.println(F("halted..."));
                    for(;;) {
                        delay(1000);
                    }
                    break;
                case '?':
                    help();
                    break;
                default:
                    break;
            }
            #if NEOPIXEL_HAVE_STATS
                const auto &stats = pixels.getStats();
                Serial.printf_P(PSTR("selected=%s value=%u brightness=%u %s dropped=%u (%.1f%%) frames=%u fps=%u delay=%u\n"),
                    get_type_str(), *valuePtr, brightness, color.toString().c_str(),
                    stats.getAbortedFrames(),  stats.getAbortedFrames() * 100.0 / stats.getFrames(), stats.getFrames(),
                    stats.getFps(), delayMillis
                );
            #else
                Serial.printf_P(PSTR("selected=%s value=%u brightness=%u %s delay=%u\n"), get_type_str(), *valuePtr, brightness, color.toString().c_str(), delayMillis);
            #endif
        }
    }

    delay(delayMillis);
    counter++;
}
